lib/pbench/test/unit/server/test_user_create.py

```diff
@@ -54,74 +61,125 @@ def test_args(server_config, pytestconfig):
+            )
+            assert not result.stderr_bytes
 
     @staticmethod
     @responses.activate
-    def test_admin_user_creation(server_config, pytestconfig):
-        runner = CliRunner(mix_stderr=False)
-        result = runner.invoke(
-            user_create.main,
-            args=[
-                TestCreateUser.USER_SWITCH,
-                TestCreateUser.USER_TEXT,
-                TestCreateUser.EMAIL_SWITCH,
-                TestCreateUser.EMAIL_TEXT,
-                TestCreateUser.FIRST_NAME_SWITCH,
-                TestCreateUser.FIRST_NAME_TEXT,
-                TestCreateUser.LAST_NAME_SWITCH,
-                TestCreateUser.LAST_NAME_TEXT,
-                TestCreateUser.ROLE_SWITCH,
-                "ADMIN",
-            ],
-            input=f"{TestCreateUser.PSWD_TEXT}\n",
-        )
-        assert result.exit_code == 0
-        assert (
-            result.stdout
-            == f"{TestCreateUser.PSWD_PROMPT}\n" + "Admin user test_user registered\n"
-        )
-        assert not result.stderr_bytes
+    def test_valid_user_delete(client, server_config, pytestconfig):
+        with client:
+            runner = CliRunner(mix_stderr=False)
+            TestUserManagement.test_args(server_config, pytestconfig)
+
+            result = runner.invoke(user_delete, args=[TestUserManagement.USER_TEXT,])
+            assert result.exit_code == 0
+            assert result.stdout == "User test_user deleted\n"
+            assert not result.stderr_bytes
 
     @staticmethod
     @responses.activate
-    def test_user_creation_with_invalid_role(server_config, pytestconfig):
-        runner = CliRunner(mix_stderr=False)
-        result = runner.invoke(
-            user_create.main,
-            args=[
-                TestCreateUser.USER_SWITCH,
-                TestCreateUser.USER_TEXT,
-                TestCreateUser.EMAIL_SWITCH,
-                TestCreateUser.EMAIL_TEXT,
-                TestCreateUser.FIRST_NAME_SWITCH,
-                TestCreateUser.FIRST_NAME_TEXT,
-                TestCreateUser.LAST_NAME_SWITCH,
-                TestCreateUser.LAST_NAME_TEXT,
-                TestCreateUser.ROLE_SWITCH,
-                "ADMN",
-            ],
-            input=f"{TestCreateUser.PSWD_TEXT}\n",
-        )
-        assert result.exit_code == 2
+    def test_valid_users_list(client, server_config, pytestconfig):
+        with client:
+            runner = CliRunner(mix_stderr=False)
+            TestUserManagement.test_args(server_config, pytestconfig)
+
+            result = runner.invoke(user_list,)
+            assert result.exit_code == 0
+            assert result.stdout == "{0:15}\t{1:15}\t{2:15}\t{3:15}\t{4:20}".format(
+                "Username", "First Name", "Last Name", "Registered On", "Email"
+            ) + "\n{0:15}\t{1:15}\t{2:15}\t{3:15}\t{4:20}\n".format(
+                "test_user", "First", "Last", "2021-04-16", "test@domain.com"
```

Comment:
I've actually wondered if we should build a yield-based fixture that sets up the database and then clears it after each test case so that test cases are independent. I suspect this would yield more durable tests... which also would in many cases be easier to write as we don't need to invent unique DB column values for each test case.

Reply:
It's easy enough for each unit test to add (or, as you say, mock) the entries it expects to list; and that way the test stands alone without depending on the side effects of other tests.

But right now, unit tests touch an actual sqlite3 DB; they're not mocked, and changes are persistent across test cases.

Reply:
My head is spinning.  How is this database created and maintained?  Is there some part of the unit test framework which does that?

Reply:
@dbutenhof the test cases are actually independent and the changes are not persisted across the test cases. We do that by defining a `pytest.fixture` client (it is defined in server `conftest.py` file `lib/pbench/test/unit/server/conftest.py`). We begin the test case by `with client: ` and once we are outside the `with` loop everything is cleared because of `setup` pytest fixture defined in the same file with `scope=session` parameter.  

So yeah @webbnh in a way the unit tests are not actually unit tests, they are more of functional tests where we spin up an actual in-memory `sqllite` database and perform all the _unit tests like operations_. However, this test gives us more real world test scenarios. These unit tests are also independent, in-memory database is cleared after each unit test and then get repopulated according to the need from 2nd unit test then gets cleared and so on.

Reply:
> These unit tests are also independent, in-memory database is cleared after each unit test and then get repopulated according to the need from 2nd unit test then gets cleared and so on.

I don't believe this is true; or, if it is for you, there's something different in my environment because I've had to choose unique naming across unit tests or suffer failures in the uniqueness checks.

Reply:
are you referring to `test_state_tracker.py` file? You have to use the `client` parameter

Reply:
Ah; I'd missed before that you're using `with` inside your test cases rather than turning the fixture into an automatic wrapper with `yield` -- which would be cleaner, more transparent, and remove the extra indentation from the test cases. But good to know.